FlowCrypt/Controllers/Inbox/InboxViewController+TableView.swift

```diff
@@ -341,55 +360,103 @@ extension InboxViewController {
-            showAlert(message: error.errorMessage)
-            return
+    func perform(action: MessageAction, inboxItems: [InboxItem]) {
+        // For permanently delete, a distinct mechanism is required. An alert must be displayed first,
+        // followed by iterating through the selected threads for deletion.
+        if action == .permanentlyDelete {
+            performPermanentDeleteAction(on: inboxItems)
+        } else {
+            performOtherAction(on: inboxItems, action: action)
         }
+    }
+
+    private func handleOperation(inboxItem: InboxItem, action: MessageAction, resetThreadSelect: Bool = true) {
+        hideSpinner()
 
         guard let indexToUpdate = getUpdatedIndex(for: inboxItem) else {
             return
         }
 
         switch action {
         case .markAsRead:
-            updateMessage(isRead: true, at: indexToUpdate)
+            updateMessage(isRead: true, at: indexToUpdate, resetThreadSelect: resetThreadSelect)
         case .markAsUnread:
-            updateMessage(isRead: false, at: indexToUpdate)
+            updateMessage(isRead: false, at: indexToUpdate, resetThreadSelect: resetThreadSelect)
         case .moveToTrash, .permanentlyDelete:
             removeMessage(at: indexToUpdate)
         case .archive, .moveToInbox:
             if path.isEmpty { // no need to remove in 'All Mail' folder
                 updateMessage(
                     labelsToAdd: action == .moveToInbox ? [.inbox] : [],
                     labelsToRemove: action == .archive ? [.inbox] : [],
-                    at: indexToUpdate
+                    at: indexToUpdate,
+                    resetThreadSelect: resetThreadSelect
                 )
             } else {
                 removeMessage(at: indexToUpdate)
             }
         }
     }
 }
+
+extension InboxViewController: InboxCellNodeDelegate {
+    private func updateInboxSelection(row: Int, isSelected: Bool) {
+        var adjustedRow = row
+        if shouldShowEmptyView { // Adjust the row index for views that show an empty row
+            adjustedRow -= 1
+        }
+        inboxInput[adjustedRow].isSelected = isSelected
+    }
+
+    @objc func handleLongPressGesture(_ recognizer: UILongPressGestureRecognizer) {
+        guard recognizer.state == .began,
+              let indexPath = tableNode.indexPathForRow(at: recognizer.location(in: tableNode.view)),
+              let inboxCellNode = tableNode.nodeForRow(at: indexPath) as? InboxCellNode else { return }
+        inboxCellNode.isCellSelected.toggle()
+        updateInboxSelection(row: indexPath.row, isSelected: inboxCellNode.isCellSelected)
+        inboxCellNode.toggleCheckBox()
+        updateNavigationBar()
+    }
+
+    func inboxCellNodeDidToggleSelection(_ node: InboxCellNode, isSelected: Bool) {
+        updateInboxSelection(row: node.indexPath!.row, isSelected: isSelected)
+        node.isCellSelected = isSelected
```

Comment:
can we use built-in `node.isSelected` for tracking node selection state?

Reply:
https://github.com/FlowCrypt/flowcrypt-ios/blob/5b21a97ce4a6c2782a9aaf953c93bab30aa2b752/FlowCryptUI/Cell%20Nodes/InboxCellNode.swift#L60

Please check this comment